File: src/main/java/org/frc1778/freezylib/logging/FreezyLog.java
package org.frc1778.freezylib.logging;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import edu.wpi.first.wpilibj.DriverStation.MatchType;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.io.Writer;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.stream.Collectors;

/**
 * A logging framework to log data from the robot to files on the robot.
 *
 * @author FRC 1778 Chill Out
 */
public class FreezyLog {

  private static List<Field> fields = new ArrayList<>();

  private static String csvFileName = "log.csv";
  private static String jsonFileName = "meta.json";

  private static String pathToLogDirectory = "/home/lvuser/logs";
  private static File csvFile;
  private static File jsonFile;

  private static boolean filesDirty = true;

  private static Gson gson =
      new GsonBuilder()
          .setPrettyPrinting()
          .serializeNulls()
          .excludeFieldsWithoutExposeAnnotation()
          .create();

  public static void setPath(String path) {
    pathToLogDirectory = path;
  }

  public static void populateMatchStructure(MatchType matchType, int matchNumber) {
    var path =
        Path.of(
            pathToLogDirectory,
            matchType.name().toLowerCase(Locale.ENGLISH),
            "match" + String.format("%02d", matchNumber));
    try {
      Files.createDirectories(path);
      jsonFile = new File(path.toString(), jsonFileName);
      csvFile = new File(path.toString(), csvFileName);
    } catch (IOException e) {
      System.err.println(e.getMessage());
    }

    filesDirty = true;
  }

  public static void setLogFileName(String name) throws Exception {
    var i = name.lastIndexOf('.');

    if (!new String(i > 0 ? name.substring(i + 1) : "").equals("csv")) {
      throw new Exception(
          "Invalid file extension, should be 'csv', but was '"
              + (i > 0 ? name.substring(i + 1) : "")
              + "'.");
    } else {
      csvFileName = name;
    }

    filesDirty = true;
  }

  public static void setMetaFileName(String name) throws Exception {
    var i = name.lastIndexOf('.');

    if (!new String(i > 0 ? name.substring(i + 1) : "").equals("json")) {
      throw new Exception(
          "Invalid file extension, should be 'json', but was '"
              + (i > 0 ? name.substring(i + 1) : "")
              + "'.");
    } else {
      jsonFileName = name;
    }

    filesDirty = true;
  }

  public static File getMetaFile() {
    return jsonFile;
  }

  public static File getLogFile() {
    return csvFile;
  }

  public static void addField(Field field) {
    fields.add(field);

    filesDirty = true;
  }

  public static void removeField(Field field) {
    fields.remove(field);

    filesDirty = true;
  }

  public static void dump() {
    fields.clear();

    filesDirty = true;
  }

  private static void pollFields() {
    fields.stream()
        .filter(o -> o instanceof PolledField)
        .map(o -> (PolledField) o)
        .forEach(PolledField::pollSupplier);
  }

  public static List<Field> getFields() {
    return fields;
  }

  public static Field getFieldByName(String name) {
    return fields.stream().filter(o -> o.getName().equals(name)).findFirst().get();
  }

  public static String getCsvHeader() {
    return fields.stream()
        .filter(o -> !o.getClass().equals(MetaField.class))
        .map(o -> o.getName())
        .collect(Collectors.joining(","));
  }

  public static String getCollectedFields() {
    return fields.stream()
        .filter(o -> !o.getClass().equals(MetaField.class))
        .map(o -> o.getValue())
        .collect(Collectors.joining(","));
  }

  public static void log() {
    pollFields();

    try {
      var csvOut = "";
      if (filesDirty) {
        csvFile.delete();
        csvOut += getCsvHeader() + "\n";

        try (Writer writer = new FileWriter(jsonFile, Charset.forName("UTF-8"), false)) {
          gson.toJson(fields, writer);
        }

        filesDirty = false;
      }

      var collectedFields = getCollectedFields();
      csvOut += collectedFields.replaceAll(",", "").length() > 0 ? collectedFields : "";

      if (csvOut.length() > 0) {
        try (PrintWriter csvWriter =
            new PrintWriter(new FileWriter(csvFile, Charset.forName("UTF-8"), true))) {
          csvWriter.write(csvOut + "\n");
        }
      }
    } catch (IOException e) {
      System.err.println(e.getMessage());
    }
  }
}

File: src/main/kotlin/org/frc1778/freezylib/logging/FreezyLog.kt
package org.frc1778.freezylib.logging

import com.google.gson.GsonBuilder
import edu.wpi.first.wpilibj.DriverStation.MatchType
import java.io.File
import java.io.IOException
import java.io.PrintWriter
import java.io.StringWriter
import java.nio.file.Files
import java.nio.file.Path
import java.util.Locale
import java.util.stream.Collectors
import org.frc1778.freezylib.util.Measurement

/**
 * A logging framework to log data from the robot to files on the robot.
 *
 * @author FRC 1778 Chill Out
 */
object FreezyLog {

    val exceptionsField = TaggedField("Exceptions", Measurement.Unitless.UNITLESS, String::class)
    private val exceptionStringWriter = StringWriter()
    private val exceptionPrintWriter = PrintWriter(exceptionStringWriter)

    private var fields = mutableListOf<Field>(exceptionsField)

    private var csvFileName = "log.csv"
    private var jsonFileName = "meta.json"

    private var pathToLogDirectory = "/home/lvuser/logs"
    lateinit var logFile: File
    lateinit var metaFile: File

    private var filesDirty = true

    private val gson =
        GsonBuilder().setPrettyPrinting().serializeNulls().excludeFieldsWithoutExposeAnnotation().create()

    val csvHeader: String
        get() = fields.stream().filter { it::class != MetaField::class }.map { o -> o.name }.collect(
            Collectors.joining(
                ","
            )
        )

    private val collectedFields: String
        get() {
            val out = fields.stream().filter { it::class != MetaField::class }.map { o -> o.value }.collect(
                Collectors.joining(
                    ","
                )
            )
            fields.stream().filter { o -> o is TaggedField }.map { o -> o as TaggedField }.forEach { it.acknowledgeTag() }
            return out
        }

    fun setPath(path: String) {
        pathToLogDirectory = path
        filesDirty = true
    }

    fun populateMatchStructure(matchType: MatchType, matchNumber: Int) {
        val path = Path.of(
                pathToLogDirectory,
                matchType.name.toLowerCase(Locale.ENGLISH),
                "match" + String.format("%03d", matchNumber))
        try {
            Files.createDirectories(path)
            metaFile = File(path.toString(), jsonFileName)
            logFile = File(path.toString(), csvFileName)
        } catch (e: IOException) {
            System.err.println(e.message)
        }

        filesDirty = true
    }

    @Throws(IOException::class)
    fun setLogFileName(name: String) {
        val i = name.lastIndexOf('.')

        if ((if (i > 0) name.substring(i + 1) else "") != "csv") {
            throw IOException(
                    "Invalid file extension, should be 'csv', but was '" +
                            (if (i > 0) name.substring(i + 1) else "") +
                            "'.")
        } else {
            csvFileName = name
        }

        filesDirty = true
    }

    @Throws(IOException::class)
    fun setMetaFileName(name: String) {
        val i = name.lastIndexOf('.')

        if ((if (i > 0) name.substring(i + 1) else "") != "json") {
            throw IOException(
                    "Invalid file extension, should be 'json', but was '" +
                            (if (i > 0) name.substring(i + 1) else "") +
                            "'.")
        } else {
            jsonFileName = name
        }

        filesDirty = true
    }

    fun addField(field: Field) {
        fields.add(field)
        filesDirty = true
    }

    fun removeField(field: Field) {
        fields.remove(field)
        filesDirty = true
    }

    fun dump() {
        fields.clear()
        addField(exceptionsField)
    }

    private fun pollFields() {
        fields.stream()
                .filter { o -> o is PolledField<*> }
                .map { o -> o as PolledField<*> }
                .forEach { it.pollSupplier() }
    }

    fun getFieldByName(name: String): Field {
        return fields.stream().filter { o -> o.name == name }.findFirst().get()
    }

    fun logException(exception: Exception) {
        exception.printStackTrace(exceptionPrintWriter)
        exceptionsField.updateTag("\"${exceptionStringWriter}\"")
    }

    fun log() {
        pollFields()

        try {
            var csvOut = ""
            if (filesDirty) {
                logFile.writeText("")
                csvOut += csvHeader + "\n"

                metaFile.writeText(gson.toJson(fields))

                filesDirty = false
            }

            val csv = collectedFields
            csvOut += if (csv.isNotEmpty()) "$csv\n" else ""
            if (csvOut.isNotEmpty()) {
                logFile.appendText("$csvOut")
            }
        } catch (e: IOException) {
            System.err.println(e.message)
        }
    }
}
